request-manager/spec/node-check.go

```diff
@@ -0,0 +1,498 @@
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ArgsNotRenamedNodeCheck struct{}
+
+/* A single arg cannot be renamed and used as multiple inputs to a node. */
+func (check ArgsNotRenamedNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	aliases := map[string]string{}
+	values := map[string]bool{}
+	for _, nodeSet := range node.Args {
+		if nodeSet.Given == nil || nodeSet.Expected == nil {
+			continue
+		}
+		if alias, ok := aliases[*nodeSet.Given]; ok && alias != *nodeSet.Expected {
+			values[*nodeSet.Given] = true
+		}
+		aliases[*nodeSet.Given] = *nodeSet.Expected
+	}
+
+	if len(values) > 0 {
+		return DuplicateValueError{sequenceName, &node.Name, "args -> given", stringSetToArray(values), "note that if `given` is not explicitly specified, then its value is the same as `expected`"}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ArgsEachOnceNodeCheck struct{}
+
+/* Each cannot alis to job arg listed in `args`. */
+func (check ArgsEachOnceNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	seen := map[string]bool{}
+	for _, nodeSet := range node.Args {
+		if nodeSet.Expected == nil {
+			continue
+		}
+		seen[*nodeSet.Expected] = true
+	}
+
+	values := map[string]bool{}
+	for _, each := range node.Each {
+		p := strings.Split(each, ":")
+		if len(p) != 2 {
+			continue
+		}
+		if seen[p[1]] {
+			values[p[1]] = true
+		}
+		seen[p[1]] = true
+	}
+
+	if len(values) > 0 {
+		return DuplicateValueError{sequenceName, &node.Name, "each", stringSetToArray(values), "in `each: arg:alias`, a given `alias` should not be repeated in `args -> expected`"}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ArgsEachNotRenamedNodeCheck struct{}
+
+/* Each cannot rename job args listed in `args`. */
+func (check ArgsEachNotRenamedNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	aliases := map[string]string{}
+	for _, nodeSet := range node.Args {
+		if nodeSet.Given == nil || nodeSet.Expected == nil {
+			continue
+		}
+		// This is fine:
+		// each: [arg:alias]
+		// args:
+		//   - expected: arg
+		if *nodeSet.Given == *nodeSet.Expected {
+			continue
+		}
+		aliases[*nodeSet.Given] = *nodeSet.Expected
+	}
+
+	values := map[string]bool{}
+	for _, each := range node.Each {
+		p := strings.Split(each, ":")
+		if len(p) != 2 {
+			continue
+		}
+		if alias, ok := aliases[p[0]]; ok && alias != p[1] {
+			values[p[0]] = true
+		}
+		aliases[p[0]] = p[1]
+	}
+
+	if len(values) > 0 {
+		return DuplicateValueError{sequenceName, &node.Name, "each", stringSetToArray(values), "in `each: arg:alias`, a given `as` value should not be repeated in `args -> given`; note that if `given` is not explicitly specified, then its value is the same as `expected`"}
+	}
+
+	return nil
+}
+
```

Comment:
After looking through all the checks related to `each`, I think maybe we should change terminology: instead of `each: arg:alias` do `each: iterable:iterator` (or `each: list:item` perhaps, I'm open to other suggestions).

Using `arg` in the first one is overloading the meaning of "arg", which is already used when talking about the `args` section of the node.